### DISCUSSION TITLE
Best way to handle non-standard OAuth2 provider

### DISCUSSION THREAD

**Discussion by leeoocca** [pocketbase/pocketbase#6016] [Q&A] [answered]
Hello, I'm trying to connect a non-standard OAuth2 provider with the generic OIDC client offered by PocketBase. By non-standard I mean that it's not OIDC-compliant, it simply supports OAuth2.

1. I found how to specify scopes in client but I'm now having issues with the user creation. As the profile endpoint of the provider is non-standard, PocketBase fails to get the right info to be used when creating the user. How would it be best I implement this logic? Should I use a custom pb_hook?

2. I'm currently getting a strange error: `providerId can't be empty`. But I'm sure it's being sent by the client in the request. How can I debug this further? Running PocketBase with `--dev` didn't help so far. I'll attach a better log tomorrow.

Thanks in advance for the great work on this project 🙏🏼

**Comment by pnmcosta** [comment, marked as answer]
1. Yes it might be possible to use a hook, but really does depend on what the requirements are, if you need to tap into the token exchange it might be tricky, but you are able to do custom user data mapping on this event https://pocketbase.io/docs/go-event-hooks/#onrecordauthwithoauth2request

2. `providerId` should be the unique identifier for the user from the auth provider, sounds like a mapping issue on your implementation. [here](https://github.com/pocketbase/pocketbase/blob/master/apis/record_auth_with_oauth2.go#L86) is the default token logic, and the default oidc [here](https://github.com/pocketbase/pocketbase/blob/master/tools/auth/oidc.go#L62) for the user mapping.

In the event I mentioned on 1. you should be able to access the default `AuthUser` mapping as well as the `RawUser` map and fill your User record with whatever is required.

**Reply by leeoocca** [reply]
Thanks a lot, Pedro!

1. After some more searching yesterday evening I bumped up on exactly that! I think I'll go this way. I don't need to do anything manually with the token, it probably simply needs the mapping to be manually specified.
2. Thanks a lot for the info. It seems to arise from the mapping error that should be fixed with point 1.

**Reply by pnmcosta** [reply]
No worries, happy to help.

**Reply by leeoocca** [reply]
So... I managed to override the creation but now I'm having issues with the login. The first time a user authenticates it creates correctly the record. I did this by extending both the `e.createData` and the `e.oAuth2User` objects. But it fails on subsequent authentications complaining that `data:email: Value must be unique. raw:email: Value must be unique.`

Do you have any suggestions?

**Reply by pnmcosta** [reply]
Umm, sounds like whatever you're doing is trying to create a new record again, email IIRC is uniquely constrained by default on auth collections, so PB and/or your hook's code is trying to create the record again.

On your first create, is the collection record and the `_externalAuth` created correctly indeed? You should be able to check this via the admin now.

**Reply by leeoocca** [reply]
Yes, the entry in `_externalAuth` is created correctly (or so I think, under `providerId` I can see the OAuth user's email).

At every successive login, `onRecordAuthWithOAuth2Request`'s event `isNewRecord` is always `true`.

**Reply by leeoocca** [reply]
Hey Pedro, thanks a lot for you support so far! Any idea how I can further debug this behaviour? It seems to be trying to create a new record every time.

**Reply by leeoocca** [reply]
Oh right, I just realised I had forgotten this sentence [from the docs](https://pocketbase.io/docs/go-event-hooks/#onrecordauthwithoauth2request):

> If e.Record is not set, then the OAuth2 request will try to create a new auth record.
> To assign or link a different existing record model you can change the e.Record field.

But what should I set to `e.record`? The `id` of the existing record? How can I quer y the db from the hook?

**Reply by pnmcosta** [reply]
> (or so I think, under providerId I can see the OAuth user's email)

Not sure this is correct, but only you can find out from your provider, I have never seen email being a "user" identifier, cause emails can change.

> But what should I set to e.record? 

You shouldn't need to explicitly set it, Pocketbase should be doing it for you if the Record auth has been properly created in the first place, and auth occurs again. See https://github.com/pocketbase/pocketbase/blob/master/apis/record_auth_with_oauth2.go#L92 you can step through the code and figure out what it's doing.

I don't think I can help much from here TBF, other than if you share a repo where the issue is easily reproducible and what the provider is and its documentation, I then may be able to help further, but this is not a promise 😅

**Reply by leeoocca** [reply]
Thanks again Pedro for all your support so far! 🙏🏼 

> Not sure this is correct, but only you can find out from your provider, I have never seen email being a "user" identifier, cause emails can change.

After I manually map the `oAuth2User` and the `createData` to use the id from the provider, `externalAuth` is created correctly with the id instead of the email. The issue is that PocketBase doesn't seem to have it when it searches the `externalAuth`. I opened a new discussion for more visibility: https://github.com/pocketbase/pocketbase/discussions/6294#discussion-7824294 

The mapping by PocketBase should work: in the profile endpoint response the id should be mapped correctly by this line

https://github.com/pocketbase/pocketbase/blob/231769501177e5e76b37bb2d0ace0ef35c8096dc/tools/auth/auth.go#L133

But then it's missing in the query...

```sql
SELECT `_externalAuths`.* FROM `_externalAuths` WHERE (NOT (`providerId`='')) AND (`collectionRef`='pbc_2651147324' AND `provider`='oidc' AND `providerId`='') ORDER BY `created` DESC LIMIT 1
```

I'm not sure how to debug this further.

> I then may be able to help further, but this is not a promise 😅

Thanks again for all the help, I'll try to build something quickly so you can see what I mean

**Reply by leeoocca** [reply]
Oooh, I think I found my issue: since I'm using the `oidc` provider, it's expecting to extract the `providerId` from the `sub` attribute in the user profile

https://github.com/pocketbase/pocketbase/blob/231769501177e5e76b37bb2d0ace0ef35c8096dc/tools/auth/oidc.go#L73

This obviously doesn't work as the provider is only oAuth2, not OIDC compliant. The id is stored in `id` and not `sub`.

If I see this correctly I only have two choices:

- To override this, I guess I have to write a custom provider and then build the binary myself, as this can't be overridden via a hook (because it's called later)
- Move to an OIDC-compliant provider

Am I right?

**Reply by pnmcosta** [reply]
I'm glad you found it 💪

Yup those 2 options would be fine. You can also propose this new provider to be added to the code base, as long as it's common enough and brings value to PB, I'm sure @ganigeorgiev would consider it.

**Reply by leeoocca** [reply]
I just ended up writing a small script that forwards the user info request to the original provider and returns `sub` as a string. Simple enough for now. I might improve it _in the future™_.

**Reply by pnmcosta** [reply]
hehehe, it's good if it works right 😅

Another thing to consider and I forgot to mention before, is the use a identity provider to handle that custom provider.

I've been looking at https://goauthentik.io/ not sure if it will do exactly what you need, but might be worth exploring.

**Reply by leeoocca** [reply]
Yes, I was looking into exactly the same thing. Unfortunately most identity providers only support OIDC themselves (as relying party) so I had no luck with that either...

I had the possibility to connect Authentik to LDAP but that also wasn't working as expected. The current solution seems the only easy one for now. Unless we have some news about https://github.com/pocketbase/pocketbase/discussions/5125

Thanks again Pedro 🙏🏼
